### PR TITLE
Relax affine

### PR DESCRIPTION
**Pull request by josel-amd**

### CODE REVIEW

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 1]
```diff
@@ -2804,7 +2806,7 @@ LogicalResult AffineIfOp::verify() {
 
   // Verify that the operands are valid dimension/symbols.
   if (failed(verifyDimAndSymbolIdentifiers(*this, getOperands(),
-                                           condition.getNumDims())))
+                                           condition.getNumDims(), true)))
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:2809]:
```suggestion
                                           condition.getNumDims(), /*allowNonAffineDimOperands=*/true)))
```

File: mlir/test/Dialect/Affine/invalid.mlir [hunk 2]
```diff
@@ -97,7 +97,6 @@ func.func @affine_if_invalid_dim(%arg : index) {
   affine.for %n0 = 0 to 7 {
     %dim = arith.addi %arg, %arg : index
```
Comment (mgehre-amd) [inline mlir/test/Dialect/Affine/invalid.mlir:99]: Can you leave a comment here, saying that we made this legal?

**Review by mgehre-amd — APPROVED** [review]

**Review by cferry-AMD — APPROVED** [review]
We can introduce stuff that breaks affine semantics by doing that, so we should be very careful. We may need to revisit this with an analysis that checks which ops generated a dimension (so `affine.for` and `affine.parallel` iterators would be legal for instance, along with `linalg.index`, and at a later point also check for affinity of expressions to allow linear combinations of indices).

**Comment by josel-amd** [comment]
> We can introduce stuff that breaks affine semantics by doing that, so we should be very careful. We may need to revisit this with an analysis that checks which ops generated a dimension (so `affine.for` and `affine.parallel` iterators would be legal for instance, along with `linalg.index`, and at a later point also check for affinity of expressions to allow linear combinations of indices).

I agree. I can imagine that this only works as long as you only do very limited things in the realm of the affine dialect, like lowering to another dialect. Other kinds of analysis and transformations might e.g. try to get the origin of the `index` and fail to infer the affinity of the value in the context of affine expressions. I hope this is only temporary until either 1) upstream properly fixes it or 2) we find a better way to achieve our goals as time goes by.

**Comment by josel-amd** [comment]
In the spirit of what I said above, I already tried to contain the changes as much as possible by only modifying the affine.if and by making sure that I'm not skipping any other checks were there before.

**Comment by cferry-AMD** [comment]
> I hope this is only temporary until either 1) upstream properly fixes it or 2) we find a better way to achieve our goals as time goes by.

It would take some time but upstream has made some moves to disconnect Linalg from affine (e.g. with the tiling interface: there are operations that are loops but have neither an affine schedule nor affine data access patterns, but can still be tiled). Something like an "affinity preserving" interface would then make sense.

**Pull request merged**